<a href="https://colab.research.google.com/github/LambdaTheda/DS-Unit-2-Applied-Modeling/blob/master/tues_APR_28_PT5_pokemon_build_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### **LOAD DATASETS**

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [4]:
# read in dataset 1: pokemon Go character sightings

import pandas as pd

df = pd.read_csv('/content/drive/My Drive/predictemall/300k.csv')

#df = pd.read_csv('300k.csv')
pd.set_option("display.max_rows", 91835)
pd.set_option("display.max_columns", 300)


/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (49) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [5]:
import pandas as pd

# read in dataset 2: encoded names and corresponding id#s
df_Names = pd.read_csv('/content/drive/My Drive/predictemall/pokemonNumbers.csv', header= None) #RT CLK ON FILE TO PUT PATH HERE OK
pd.set_option("display.max_rows", 91835)
df_Names.columns=['pokemonId', 'pokemonName'] # change df_Names column names to align with the proper columns for merge.

df_Names

pokemonId pokemonName
0            1   Bulbasaur
1            2     Ivysaur
2            3    Venusaur
3            4  Charmander
4            5  Charmeleon
5            6   Charizard
6            7    Squirtle
7            8   Wartortle
8            9   Blastoise
9           10    Caterpie
10          11     Metapod
11          12  Butterfree
12          13      Weedle
13          14      Kakuna
14          15    Beedrill
15          16      Pidgey
16          17   Pidgeotto
17          18     Pidgeot
18          19     Rattata
19          20    Raticate
20          21     Spearow
21          22      Fearow
22          23       Ekans
23          24       Arbok
24          25     Pikachu
25          26      Raichu
26          27   Sandshrew
27          28   Sandslash
28          29   Nidoran F
29          30    Nidorina
30          31   Nidoqueen
31          32   Nidoran M
32          33    Nidorino
33          34    Nidoking
34          35    Clefairy
35          36    Clefable
36          37      Vulpix
37          38   Ninetales
38          39  Jigglypuff
39          40  Wigglytuff
40          41       Zubat
41          42      Golbat
42          43      Oddish
43          44       Gloom
44          45   Vileplume
45          46       Paras
46          47    Parasect
47          48     Venonat
48          49    Venomoth
49          50     Diglett
50          51     Dugtrio
51          52      Meowth
52          53     Persian
53          54     Psyduck
54          55     Golduck
55          56      Mankey
56          57    Primeape
57          58   Growlithe
58          59    Arcanine
59          60     Poliwag
60          61   Poliwhirl
61          62   Poliwrath
62          63        Abra
63          64     Kadabra
64          65    Alakazam
65          66      Machop
66          67     Machoke
67          68     Machamp
68          69  Bellsprout
69          70  Weepinbell
70          71  Victreebel
71          72   Tentacool
72          73  Tentacruel
73          74     Geodude
74          75    Graveler
75          76       Golem
76          77      Ponyta
77          78    Rapidash
78          79    Slowpoke
79          80     Slowbro
80          81   Magnemite
81          82    Magneton
82          83  Farfetch'd
83          84       Doduo
84          85      Dodrio
85          86        Seel
86          87     Dewgong
87          88      Grimer
88          89         Muk
89          90    Shellder
90          91    Cloyster
91          92      Gastly
92          93     Haunter
93          94      Gengar
94          95        Onix
95          96     Drowzee
96          97       Hypno
97          98      Krabby
98          99     Kingler
99         100     Voltorb
100        101   Electrode
101        102   Exeggcute
102        103   Exeggutor
103        104      Cubone
104        105     Marowak
105        106   Hitmonlee
106        107  Hitmonchan
107        108   Lickitung
108        109     Koffing
109        110     Weezing
110        111     Rhyhorn
111        112      Rhydon
112        113     Chansey
113        114     Tangela
114        115  Kangaskhan
115        116      Horsea
116        117      Seadra
117        118     Goldeen
118        119     Seaking
119        120      Staryu
120        121     Starmie
121        122    Mr. Mime
122        123     Scyther
123        124        Jynx
124        125  Electabuzz
125        126      Magmar
126        127      Pinsir
127        128      Tauros
128        129    Magikarp
129        130    Gyarados
130        131      Lapras
131        132       Ditto
132        133       Eevee
133        134    Vaporeon
134        135     Jolteon
135        136     Flareon
136        137     Porygon
137        138     Omanyte
138        139     Omastar
139        140      Kabuto
140        141    Kabutops
141        142  Aerodactyl
142        143     Snorlax
143        144    Articuno
144        145      Zapdos
145        146     Moltres
146        147     Dratini
147      

### **EXPLORATORY DATA ANALYSES AND WRANGLING/MUNGING/CLEANING**

In [6]:
df.head()

pokemonId   latitude   longitude    appearedLocalTime  \
0         16  20.525745  -97.460829  2016-09-08T03:57:45   
1        133  20.523695  -97.461167  2016-09-08T03:57:37   
2         16  38.903590  -77.199780  2016-09-08T03:57:25   
3         13  47.665903 -122.312561  2016-09-08T03:56:22   
4        133  47.666454 -122.311628  2016-09-08T03:56:08   

                            _id           cellId_90m          cellId_180m  \
0  NTgxMDkzOTk4MTM5MjUwMjIzNw==  9645139108510564000  9645139108711890000   
1  OTQ1NDgzODc1MjM3NDEzMTI2MQ==  9645139109852742000  9645139109785633000   
2  NTQ0OTQ0NDA1Nzg2ODg3OTg2OQ==  9923201472785285000  9923201472986612000   
3  NTU2MTU1NDM4NzA2MDk1MDcxNw==  6093392705025474600  6093392705092583400   
4  MTY2ODg4MTAzMTczMDE0MTUwNTM=  6093392707709829100  6093392707776938000   

           cellId_370m          cellId_730m         cellId_1460m  \
0  9645139108443455000  9645139109517197000  9645139113812165000   
1  9645139110590940000  9645139109517197000  9645139113812165000   
2  9923201473791918000  9923201477013144000  9923201481308110000   
3  6093392705897889800  6093392702676664300  6093392715561566200   
4  6093392708045373400  6093392711266598900  6093392715561566200   

          cellId_2920m         cellId_5850m appearedTimeOfDay  appearedHour  \
0  9645139130992034000  9645138924833604000             night             5   
1  9645139130992034000  9645138924833604000             night             5   
2  9923201498487980000  9923201567207457000             night             5   
3  6093392767101173800  6093392560942743600             night             5   
4  6093392767101173800  6093392560942743600             night             5   

   appearedMinute appearedDayOfWeek  appearedDay  appearedMonth  appearedYear  \
0              57         dummy_day            8              8          2016   
1              57         dummy_day            8              8          2016   
2              57         dummy_day            8              8          2016   
3              56         dummy_day            8              8          2016   
4              56         dummy_day            8              8          2016   

   terrainType  closeToWater         city continent       weather  \
0           14         False  Mexico_City   America         Foggy   
1           14         False  Mexico_City   America         Foggy   
2           13         False     New_York   America         Clear   
3            0          True  Los_Angeles   America  PartlyCloudy   
4            0          True  Los_Angeles   America  PartlyCloudy   

   temperature  windSpeed  windBearing  pressure          weatherIcon  \
0         25.5       4.79          269   1018.02                  fog   
1         25.5       4.79          269   1018.02                  fog   
2         24.2       4.29          218   1015.29          clear-night   
3         15.6       5.84          160   1020.52  partly-cloudy-night   
4         15.6       5.84          160   1020.52  partly-cloudy-night   

   sunriseMinutesMidnight  sunriseHour  sunriseMinute  sunriseMinutesSince  \
0                     436            7             16                  941   
1                     436            7             16                  941   
2                     404            6             44                 1033   
3                     398            6             38                  858   
4                     398            6             38                  858   

   sunsetMinutesMidnight  sunsetHour  sunsetMinute  sunsetMinutesBefore  \
0                   1181          19            41                 -196   
1                   1181          19            41                 -196   
2                   1171          19            31                 -266   
3                   1179          19            39                  -77   
4                   1179          19            39                  -77   

   population_density  urban  subu

In [7]:
df.shape

(296021, 208)

In [8]:
df.sample(100)

pokemonId   latitude   longitude    appearedLocalTime  \
119319         19  50.057654   14.502856  2016-09-05T20:11:58   
136428         98  20.560991  -97.445953  2016-09-05T13:20:12   
290402         74  34.008986 -118.496478  2016-09-03T00:29:44   
99954          58  34.276895 -119.208603  2016-09-06T02:22:29   
122012         85  44.496245   11.343951  2016-09-05T19:24:54   
133137         69  52.877546   18.798514  2016-09-05T14:03:20   
169880         16  40.058999  -84.246442  2016-09-05T02:34:21   
67253          41  50.677029   21.749485  2016-09-06T13:12:32   
128785         19  45.463331    9.176435  2016-09-05T17:08:52   
283873         20  39.315312  -84.523652  2016-09-03T03:20:42   
264307         90  60.517482    4.911362  2016-09-03T09:54:54   
121038         41  44.425265   26.104487  2016-09-05T19:37:59   
200067        129  30.207201  -85.863297  2016-09-04T14:24:55   
176687         16  40.163163 -104.976991  2016-09-04T23:41:14   
171606         96  53.541899 -113.494747  2016-09-05T02:00:54   
103347        127  29.632319  -98.610012  2016-09-06T00:20:54   
116230         32  40.790291  -73.947509  2016-09-05T21:16:00   
137166         16  43.005557  -73.848385  2016-09-05T12:45:04   
253770         42  45.168322    1.560809  2016-09-03T14:24:17   
244135         16  33.504010 -117.147477  2016-09-03T18:54:49   
58647          96  43.706200  -79.402908  2016-09-06T15:51:11   
209935         13  52.153781    0.143158  2016-09-04T10:42:24   
281915         69  20.513794  -97.461167  2016-09-03T03:55:50   
173340         13  41.375453  -85.036378  2016-09-05T01:14:53   
114542         20  34.899829  -77.527591  2016-09-05T21:42:29   
137163         60   1.303941  103.810260  2016-09-05T12:45:12   
220080         46  32.825837 -116.903042  2016-09-04T06:12:40   
160274         52  39.755496 -105.225377  2016-09-05T05:42:04   
80098          46  30.421713  -91.060073  2016-09-06T09:29:10   
194162         13  50.741170   -1.836282  2016-09-04T16:16:42   
215203         16  21.573916 -158.109093  2016-09-04T08:17:37   
28565         120  30.041711  -90.236156  2016-09-07T01:58:54   
291924         10  54.724916   25.255365  2016-09-03T00:07:56   
133348        133   1.520398  110.335303  2016-09-05T14:00:10   
86592         133  49.687592   14.007287  2016-09-06T07:18:15   
233641         48  44.351621  -72.742492  2016-09-03T23:32:29   
191048         16  18.155510  -67.140340  2016-09-04T17:34:31   
29344         133  41.615080  -87.840816  2016-09-07T01:42:10   
225368        133  20.476793  -97.445868  2016-09-04T03:41:04   
279533         19  32.788883  130.746782  2016-09-03T04:56:47   
57553          16  49.837416   18.288322  2016-09-06T16:05:06   
269032         19  53.595544 -113.399570  2016-09-03T08:26:24   
270641         13  46.209090    6.147351  2016-09-03T07:55:08   
248329         10  32.606685  -97.153481  2016-09-03T16:34:51   
242509         16  49.274263   -0.454162  2016-09-03T19:15:42   
180688         46  39.518997 -104.761893  2016-09-04T22:08:43   
35014           7  30.041711  -90.236156  2016-09-06T23:58:54   
288683         13  53.644207 -113.504073  2016-09-03T01:31:16   
112035         10  45.522757    9.599365  2016-09-05T22:14:50   
215611         16  48.311172   14.286496  2016-09-04T08:11:22   
36663          14  53.683998   -0.431926  2016-09-06T23:33:00   
57104          19  35.781541  -78.639297  2016-09-06T16:10:02   
94410          13  44.337446  -72.756862  2016-09-06T04:11:04   
294740         41  34.155239 -118.163878  2016-09-02T22:22:21   
292420         13  38.497605  -77.546671  2016-09-02T23:53:44   
69655          54  52.367408   -7.710272  2016-09-06T12:37:28   
287755         16  35.735469  -78.656761  2016-09-03T01:59:17   
138313         29  49.220454   16.555872  2016-09-05T12:22:53   
161690         16   1.342070  103.682812  2016-09-05T05:12:56   
90302          16  47.476019    8.305823  2016-09-06T05:46:41   
152829         41  45.383672 -122.755652  2

In [9]:
df_Names.head(20)

pokemonId pokemonName
0           1   Bulbasaur
1           2     Ivysaur
2           3    Venusaur
3           4  Charmander
4           5  Charmeleon
5           6   Charizard
6           7    Squirtle
7           8   Wartortle
8           9   Blastoise
9          10    Caterpie
10         11     Metapod
11         12  Butterfree
12         13      Weedle
13         14      Kakuna
14         15    Beedrill
15         16      Pidgey
16         17   Pidgeotto
17         18     Pidgeot
18         19     Rattata
19         20    Raticate

In [10]:
df_Names.shape

(150, 2)

In [11]:
#df3 is MERGED df and df_Names
df3 =  pd.merge(df_Names, df, on = 'pokemonId')

import pandas as pd

pd.set_option("display.max_rows", 91835)
# merge df and df_Names to make df3
df3.head()


pokemonId pokemonName   latitude   longitude    appearedLocalTime  \
0          1   Bulbasaur  38.207112  -81.512109  2016-09-08T03:24:48   
1          1   Bulbasaur  47.796156 -122.183200  2016-09-08T03:19:33   
2          1   Bulbasaur  20.525230  -97.459899  2016-09-08T03:01:13   
3          1   Bulbasaur  32.908984  -97.323683  2016-09-08T02:28:26   
4          1   Bulbasaur  19.250677  -99.605213  2016-09-08T02:17:09   

                            _id           cellId_90m          cellId_180m  \
0  MTY1NzU4NzIxOTQ1ODE2NTQ3MDE=  9822025167100969000  9822025167302296000   
1  MTcyNzA2NjM4Mjg1NDQ1MTQ0OTM=  6093386923328405500  6093386923529732100   
2  MTY5MTc0NzAzODg0NDY2NDc2NjE=  9645139108779000000  9645139108711890000   
3  NzQyNDgwNjAzNjM3NTE2OTU5Nw==  9677629639664600000  9677629639731708000   
4  MTI5ODI4NzUyNTc0ODA1ODIwNQ==  9641516016638886000  9641516016571777000   

           cellId_370m          cellId_730m         cellId_1460m  \
0  9822025167570730000  9822025164349506000  9822025177234407000   
1  6093386922724425700  6093386921650683900  6093386908765782000   
2  9645139108443455000  9645139109517197000  9645139113812165000   
3  9677629640537014000  9677629643758240000  9677629648053207000   
4  9641516015766470000  9641516012545245000  9641516016840212000   

          cellId_2920m         cellId_5850m appearedTimeOfDay  appearedHour  \
0  9822025125694800000  9822025056975323000             night             5   
1  6093386857226174500  6093387063384604700             night             5   
2  9645139130992034000  9645138924833604000             night             5   
3  9677629699592815000  9677629493434384000             night             4   
4  9641515965300605000  9641516034020082000             night             4   

   appearedMinute appearedDayOfWeek  appearedDay  appearedMonth  appearedYear  \
0              24         dummy_day            8              8          2016   
1              19         dummy_day            8              8          2016   
2               1         dummy_day            8              8          2016   
3              28         dummy_day            8              8          2016   
4              17         dummy_day            8              8          2016   

   terrainType  closeToWater         city continent       weather  \
0            4         False     New_York   America         Clear   
1            1         False  Los_Angeles   America  PartlyCloudy   
2           14         False  Mexico_City   America         Foggy   
3           10         False      Chicago   America         Clear   
4           13         False  Mexico_City   America  MostlyCloudy   

   temperature  windSpeed  windBearing  pressure          weatherIcon  \
0         21.6       4.23          190   1018.39          clear-night   
1         15.0       3.27          135   1020.57  partly-cloudy-night   
2         25.5       4.79          269   1018.02                  fog   
3         29.8       9.30          160   1014.39          clear-night   
4         14.2       5.30          178   1011.80  partly-cloudy-night   

   sunriseMinutesMidnight  sunriseHour  sunriseMinute  sunriseMinutesSince  \
0                     422            7              2                  982   
1                     397            6             37                  822   
2                     436            7             16                  941   
3                     429            7              9                  869   
4                     446            7             26                  913   

   sunsetMinutesMidnight  sunsetHour  sunsetMinute  sunsetMinutesBefore  \
0                   1188          19            48                 -216   
1                   1178          19            38                  -41   
2                   1181          19            41                 -196   
3                   1187          19            47                 -111   
4                   1189          19  

In [12]:
df3.shape

(296021, 209)

In [13]:
df3.sample(100)

pokemonId pokemonName   latitude   longitude    appearedLocalTime  \
257017        102   Exeggcute  34.899661  -77.526785  2016-09-05T04:29:51   
274622        128      Tauros  38.954015  -77.381922  2016-09-05T10:39:11   
229648         77      Ponyta  46.851704 -100.808709  2016-09-05T18:11:00   
164463         32   Nidoran M  36.863037 -119.792317  2016-09-07T03:03:20   
277093        129    Magikarp  47.649232 -122.299104  2016-09-06T08:24:47   
260433        111     Rhyhorn  39.760228 -105.012595  2016-09-04T04:13:56   
295059        140      Kabuto  33.316448 -111.976050  2016-09-06T03:27:10   
87620          16      Pidgey  50.102606   14.400483  2016-09-03T16:01:43   
5311           10    Caterpie  42.009083    2.288704  2016-09-06T20:19:31   
210423         54     Psyduck  50.065763   14.433848  2016-09-04T03:49:43   
31447          13      Weedle  44.334977  -72.751905  2016-09-04T18:18:03   
182535         41       Zubat  42.277596  -87.830146  2016-09-03T14:24:52   
66165          16      Pidgey  19.249565  -99.607127  2016-09-05T17:30:19   
181163         41       Zubat  43.925104   11.049919  2016-09-04T09:37:15   
191269         46       Paras  40.081707 -104.811217  2016-09-06T03:10:52   
228796         74     Geodude  33.246658 -111.647953  2016-09-03T04:28:46   
109656         19     Rattata  17.419700  102.785217  2016-09-06T12:24:39   
66275          16      Pidgey  20.524700  -97.459308  2016-09-05T17:18:20   
228861         74     Geodude -33.489063  -70.542065  2016-09-03T02:17:09   
166788         32   Nidoran M  26.545316  -80.041223  2016-09-04T21:25:55   
129712         19     Rattata  43.905183  -79.257933  2016-09-04T03:54:04   
180129         41       Zubat  27.317801  -97.679187  2016-09-04T22:25:10   
84388          16      Pidgey  39.006226  -77.486195  2016-09-04T01:03:42   
50361          16      Pidgey  25.724809 -100.373344  2016-09-06T23:20:45   
189853         46       Paras  32.824124  -96.714574  2016-09-06T23:22:19   
179071         41       Zubat -37.561263  143.861967  2016-09-05T08:58:33   
231937         79    Slowpoke  35.737346  -78.650822  2016-09-03T16:16:17   
7929           10    Caterpie  38.987333  -84.702368  2016-09-05T17:18:25   
103969         19     Rattata  34.900228  -77.525710  2016-09-07T00:54:31   
70399          16      Pidgey  46.103899   13.212164  2016-09-05T08:34:55   
224656         69  Bellsprout  49.338131   17.993230  2016-09-03T08:53:05   
230441         79    Slowpoke  43.767577  -79.214053  2016-09-07T21:38:56   
180413         41       Zubat  53.875171   -2.393532  2016-09-04T18:21:04   
276137        129    Magikarp  49.589213   -1.263640  2016-09-06T17:19:31   
279345        129    Magikarp  44.336076  -72.763275  2016-09-05T05:10:52   
247694         96     Drowzee  56.150468   13.604841  2016-09-03T20:55:08   
162447         29   Nidoran F  41.674631    2.791247  2016-09-04T22:43:54   
177550         41       Zubat  52.366816   -7.716765  2016-09-05T23:00:41   
223892         69  Bellsprout  49.599741   18.142599  2016-09-04T13:00:40   
293633        133       Eevee  20.525379  -97.424067  2016-09-03T07:36:22   
254474         98      Krabby  46.176465 -123.013635  2016-09-03T05:19:52   
255458        100     Voltorb  41.114209  -73.327715  2016-09-05T01:09:30   
276224        129    Magikarp  55.925373   13.541191  2016-09-06T16:19:29   
204812         49    Venomoth  59.210972    9.610925  2016-09-04T17:47:23   
40177          13      Weedle  41.526256  -87.468484  2016-09-03T06:12:15   
229541         77      Ponyta  25.726384 -100.380715  2016-09-06T01:09:23   
192152         46       Paras  46.089438   14.511893  2016-09-05T14:59:44   
178593         41       Zubat  49.844730   18.288598  2016-09-05T12:17:06   
206590         52      Meowth  30.041345  -90.239089  2016-09-05T06:00:32   
292581        133       Eevee  40.285174 -111.739341  2016-09-03T19:15:03   
205001         50     Diglett  34.001381  -81.014514  2016-09-07T00:35:42   
2181

In [14]:
# Choosing'appearedDayOfWeek' as target
df3['appearedDayOfWeek'].nunique()

7

In [17]:
# Peek at 'appearedCity' as possible better target ('dummy_day' taken as 'Monday,' but dubious, based on  discussions, metadata in Kaggle)
df3['city'].value_counts()

New_York        54177
Chicago         32046
Prague          28000
Los_Angeles     22108
London          19094
Rome            15883
Mexico_City     12672
Paris           11210
Stockholm        9786
Zurich           6563
Edmonton         6007
Oslo             5910
Denver           5619
Phoenix          5429
Ljubljana        5361
Madrid           4623
Warsaw           4369
Toronto          3613
Vancouver        3293
Tokyo            2803
Bangkok          2789
Dublin           2627
Kuching          2157
Melbourne        2113
Amsterdam        2096
Vienna           2030
Auckland         1807
Brunei           1696
Berlin           1518
Bucharest        1456
Singapore        1421
Hong_Kong        1313
Manila           1189
Buenos_Aires     1074
Sao_Paulo        1052
Indianapolis     1049
Monterrey        1010
Winnipeg          789
Regina            742
Ho_Chi_Minh       725
Sydney            676
Copenhagen        621
Puerto_Rico       616
Detroit           502
Kuala_Lumpur      488
Kolkata   

In [18]:
df3['appearedDayOfWeek'].value_counts()

Saturday     72201
Friday       68037
Wednesday    60115
Thursday     54257
Sunday       32056
dummy_day     9001
Tuesday        354
Name: appearedDayOfWeek, dtype: int64

In [0]:
# check df3 Nulls- returns 0 for all columns
df3.isnull().sum()

In [16]:
df3.columns.to_list()

['pokemonId',
 'pokemonName',
 'latitude',
 'longitude',
 'appearedLocalTime',
 '_id',
 'cellId_90m',
 'cellId_180m',
 'cellId_370m',
 'cellId_730m',
 'cellId_1460m',
 'cellId_2920m',
 'cellId_5850m',
 'appearedTimeOfDay',
 'appearedHour',
 'appearedMinute',
 'appearedDayOfWeek',
 'appearedDay',
 'appearedMonth',
 'appearedYear',
 'terrainType',
 'closeToWater',
 'city',
 'continent',
 'weather',
 'temperature',
 'windSpeed',
 'windBearing',
 'pressure',
 'weatherIcon',
 'sunriseMinutesMidnight',
 'sunriseHour',
 'sunriseMinute',
 'sunriseMinutesSince',
 'sunsetMinutesMidnight',
 'sunsetHour',
 'sunsetMinute',
 'sunsetMinutesBefore',
 'population_density',
 'urban',
 'suburban',
 'midurban',
 'rural',
 'gymDistanceKm',
 'gymIn100m',
 'gymIn250m',
 'gymIn500m',
 'gymIn1000m',
 'gymIn2500m',
 'gymIn5000m',
 'pokestopDistanceKm',
 'pokestopIn100m',
 'pokestopIn250m',
 'pokestopIn500m',
 'pokestopIn1000m',
 'pokestopIn2500m',
 'pokestopIn5000m',
 'cooc_1',
 'cooc_2',
 'cooc_3',
 'cooc_4',


### **GET BASELINE**

In [0]:
# The first way to get the highest accuracy is by assuming the predicted y_value will be of the Majority Class; the one that occurs MOST OFTEN:

pd.set_option('display.max_columns', 500)
pd.set_option('display.max_row', 300000)
'''
average for regression should be the majority value in a sense; 
classification is assuming all values are the highest majority value- BASELINE IS PREDICTING MODE CLASS (largest # of observations
'''

df3['appearedDayOfWeek'].value_counts()


#### our **BASELINE IS SATURDAY**, WITH 72201 appearances on that day

#**BUILD PREDICTION MODEL**

In [0]:
# TARGET:
df3['appearedDayOfWeek']

### **CALCULATING ACCURACY SCORE**- using df.replace() instead of encoding days of week

In [0]:
#convert days of week to Floats for computing accuracy
df3['appearedDayOfWeek'].replace('Sunday', 1.0, inplace = True)
df3['appearedDayOfWeek'].replace('dummy_day', 2.0, inplace = True)
df3['appearedDayOfWeek'].replace('Tuesday', 3.0, inplace = True)
df3['appearedDayOfWeek'].replace('Wednesday', 4.0, inplace = True)
df3['appearedDayOfWeek'].replace('Thursday', 5.0, inplace = True)
df3['appearedDayOfWeek'].replace('Friday', 6.0, inplace = True)
df3['appearedDayOfWeek'].replace('Saturday', 7.0, inplace = True)
df3.sample(100)

### **VISUALIZATION 1: CORRELATION HEATMAP**

### use correlation matrix and heatmap to help determine which features to include in models

In [0]:
#For Correlation matrix: setting categories to type 'category' for faster operations
#df['column1'] = df['column1'].astype('category') ?? = typecasting

df3.corr().isnull().sum()

In [0]:
#View correlation matrix- at the moment only interested in features most Positively or most Negatively correlated with 'appearedDayOfWeek' 
df3.corr()

In [0]:
#Correlation heatmap: https://towardsdatascience.com/better-heatmaps-and-correlation-matrix-plots-in-python-41445d0f2bec
import seaborn as sns
import matplotlib.pyplot as plt

corrmat = df3.corr()
f, ax = plt.subplots(figsize=(50, 50))
sns.heatmap(corrmat, vmax=.8, square=True);

In [0]:
#trim df to reduce noise: try Binning


In [19]:
df3['appearedDayOfWeek'].nunique()

7

In [0]:
# Fit a Classification model
# Try a shallow decision tree as a fast, first model

#pip install --upgrade category_encoders
# from sklearn.preprocessing import CategoricalEncoder
#pip install category_encoders

!pip install category_encoders==2.* 

import numpy as np
import category_encoders as ce
from sklearn.pipeline import make_pipeline
from sklearn.impute import SimpleImputer
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

#encoder = ce.OrdinalEncoder(cols=[...])

target = 'appearedDayOfWeek'
#features = ['appearedMonth', 'appearedYear', 'appearedDay', 'pokestopDistanceKm', 'weather', 'latitude', 'longitude'] #CAUSING LEAKS, 1ST 3 COLS??
features = ['pokestopDistanceKm', 'weather', 'latitude', 'longitude']

'''
#Give 2 parameters pass df to make X and y train from features, then ?...

#df3[features] eg takes first 70 rows (for 70-30 split) to make x train of all these then to 

#y_train of first 70 rows take 

#last 30 rows are Xtest and y_test are y_target (cols); default split size is 75% train, 25 test
'''
#a sklearn fcn to split df3 into test and train sets
X_train, X_test, y_train, y_test = train_test_split(df3[features], df3[target])

#train, test = train_test_split(df)- Bruno thought was synonymous with above line

pipeline = make_pipeline(
    ce.OrdinalEncoder(),
    SimpleImputer(strategy='most_frequent'),
    RandomForestClassifier() #can add parameters n-folds or k-folds.. tree depth
)


In [0]:
#make y_train #fit model 1: Random Forest Classifier and score accuracy.
'''
Model 1 Test Accuracy: 0.6300570224035889; BEATS Baseline Accuracy: 0.24476393805907629
'''

In [0]:

pipeline.fit(X_train, y_train)

#acc_score = pipeline.score(df3[features], df3[target])

#acc_score = pipeline.score(X_test, y_test) #runs pipeline ag X_Test and comps with actuals = y_test

#ra_score = roc_auc_score(test[target], pipeline.predict(test[features]))

print(f'Test Accuracy: {pipeline.score(X_test, y_test)}') # obivates #acc_score = pipeline.score(X_test, y_test) # runs pipeline ag X_Test and comps with actuals = y_test
#bc pipeline does it all

#print(f'Test ROC AUC: {roc_auc_score(y_test, pipeline.predict(X_test))}\n')


In [0]:
# fit Random Forest Classifier 2nd time, as above but with ['cellId_180m', 'cellId_370m'] as features
'''
Returns: Test Accuracy: 0.5204848255546848 - much better than 
Model 1 Test Accuracy: 0.6300570224035889; 1st model's features = ['pokestopDistanceKm', 'weather', 'latitude', 'longitude']
(both) BEAT(S) Baseline Accuracy: 0.24476393805907629
'''

target = 'appearedDayOfWeek'
features = ['cellId_180m', 'cellId_370m']

#a sklearn fcn to split df3 into test and train sets
X_train, X_test, y_train, y_test = train_test_split(df3[features], df3[target])

pipeline = make_pipeline(
    ce.OrdinalEncoder(),
    SimpleImputer(strategy='most_frequent'),
    RandomForestClassifier() #can add parameters n-folds or k-folds.. tree depth
)

pipeline.fit(X_train, y_train)

print(f'Test Accuracy: {pipeline.score(X_test, y_test)}') # obivates #acc_score = pipeline.score(X_test, y_test) # runs pipeline ag X_Test and comps with actuals = y_test


In [0]:
# fit Random Forest Classifier 3RD time, as above but with ['cellId_180m', 'cellId_370m', 'cellId_730m',	'cellId_1460m',	'cellId_2920m'] as features
'''
3rd RFC Model Returns: Test Accuracy: 0.5189173850768857- BEATS Baseline Accuracy: 0.24476393805907629
        Performed a bit WORSE than 1st model's features = ['pokestopDistanceKm', 'weather', 'latitude', 'longitude']
        Model 1 Test Accuracy: 0.6300570224035889
'''

target = 'appearedDayOfWeek'
features = ['cellId_180m', 'cellId_370m', 'cellId_730m','cellId_1460m',	'cellId_2920m']

#a sklearn fcn to split df3 into test and train sets
X_train, X_test, y_train, y_test = train_test_split(df3[features], df3[target])

pipeline = make_pipeline(
    ce.OrdinalEncoder(),
    SimpleImputer(strategy='most_frequent'),
    RandomForestClassifier() #can add parameters n-folds or k-folds.. tree depth
)

pipeline.fit(X_train, y_train)

print(f'Test Accuracy: {pipeline.score(X_test, y_test)}') # obivates #acc_score = pipeline.score(X_test, y_test) # runs pipeline ag X_Test and comps with actuals = y_test


### **CALCULATE BASELINE ACCURACY**

 
To get accuracy of Baseline for comparison-  against test set of 30% of rows

i need 30% preds = Saturday for baseline

run a loop iterate over 30% of rows from test set (y_test) against the baseline (hardcode  as 7.0) 
and chk the target val to see if  set =0 before loop,
inside loop, if the acutal val from test set is == baseline add to
 Correct Predictions   (a variable)
   else
     add to inCorrect Predictions 
     
     ie comp actual to predictions




In [0]:
#GET BASELINE (mode of 'appearedDayOfWeek' ) - getting ~25%
df3['appearedDayOfWeek'].value_counts()

In [0]:
CorrPreds=0
InCorrPreds=0

for z in y_test: 
  if z == 7.0:
     CorrPreds+=1
  else:
      InCorrPreds+=1
print(CorrPreds, InCorrPreds)     #18184 Corr preds  55822  incorr; NOW OUR BASELINE ACCURACY = 
print('Baseline Acc is :  ', (CorrPreds/(CorrPreds + InCorrPreds))) 

# Logistic Regression for Classification Model

In [0]:
# 214._logistic_regression.ipynb   https://colab.research.google.com/drive/1_ruFrbm1isf4VWTBQbkI9bzqh8LwbOwf#scrollTo=K35e_UsqOcV8
from sklearn.model_selection import train_test_split

#train, val = train_test_split(train, random_state=42) -NOT WORK
X_train, X_test, y_train, y_test = train_test_split(df3[features], df3[target])

print(X_train.shape, X_test.shape) # Changed from 11pm SUN Mar 8 build 2.ipynb; LAST W/ED ntbk

#WAS print(train.shape, val.shape)   https://colab.research.google.com/drive/18AbQ4NV9ivrYrkVMZnRIniFb0hdmvpgY#scrollTo=sOdMUOlodXzU&uniqifier=1 


In [0]:
LR_pipeline = make_pipeline(
    ce.OrdinalEncoder(),
    SimpleImputer(strategy='most_frequent'),
    LogisticRegression() #can add parameters n-folds or k-folds.. tree depth
)

In [0]:
#fit LR model

LR_pipeline.fit(X_train, y_train)

In [0]:
#Score on test set for LR Model
print(f'Test Accuracy: {LR_pipeline.score(X_test, y_test)}') # obivates #acc_score = pipeline.score(X_test, y_test) # runs pipeline ag X_Test and comps with actuals = y_test


# **Visualizations**

In [0]:
#1) see correlation matrix in previous cell

In [0]:
#2) PERMUTATION IMPORTANCES
#   a) just to peek at which features are important to our model, get feature importances
rf = pipeline.named_steps['randomforestclassifier']
importances = pd.Series(rf.feature_importances_, X_train.columns)

# Plot FEATURE IMPORTANCES
%matplotlib inline
import matplotlib.pyplot as plt

n = 4
plt.figure(figsize=(10,n/2))
plt.title(f'Top {n} features')
importances.sort_values()[-n:].plot.barh(color='grey');

In [0]:
# BEFORE PERMUTING: Sequence of the feature to be permuted: from Features Importance above,
# chose Latitude and Inspection type columns/features to Permute
import numpy as np
for feature in ['cellId_180m', 'cellId_370m']:

# PERMUTE LATITUDE AND INSPECTION TYPE FEATURES
  X_train_permuted = X_train.copy() #copy whole df to submit all at once
  X_train_permuted[feature] = np.random.permutation(X_train[feature])

  X_test_permuted = X_test.copy()
  X_test_permuted[feature] = np.random.permutation(X_test[feature])

  score = pipeline.score(X_test, y_test)
  score_permuted = pipeline.score(X_test_permuted, y_test) #Calc. accuracy on the permuted val dataset

  acc_score = pipeline.score(X_test, y_test)

  print(f'Validation accuracy with {feature}: {score}')
  print(f'Validation accuracy with {feature} permuted: {score_permuted}')
  print(f'Permutation importance: {acc_score - score_permuted}\n')

In [0]:
#3) Get feature importances
rf = pipeline.named_steps['randomforestclassifier']
pipeline.fit(X_train, y_train)
importances = pd.Series(rf.feature_importances_, X_train.columns)

# Plot feature importances
%matplotlib inline
import matplotlib.pyplot as plt

n = 4
plt.figure(figsize=(10,n/2))
plt.title(f'Top {n} features')
importances.sort_values()[-n:].plot.barh(color='grey');

In [0]:
#split df3 for val set
#df3['continent'].value_counts()

In [0]:
#2) Shapley Values: SHAP Values (an acronym from SHapley Additive exPlanations) break down a prediction to show the impact of each feature.
# from https://colab.research.google.com/drive/1r2VFMtBAt3sLVIQfsMWyQXt8hB9gziRA#scrollTo=Ep1aBVpVcrDj (FINAL VERSION 234 u2s3m4.ipynb)

import category_encoders as ce
from sklearn.impute import SimpleImputer
from sklearn.pipeline import make_pipeline
from xgboost import XGBClassifier

processor = make_pipeline(
    ce.OrdinalEncoder(), 
    SimpleImputer(strategy='median')
)

#try to split val to be 20% of df3? Or try to use test set instead of val
#val = df3[df3['continent']]
X_test = features
y_test = target

X_train_processed = processor.fit_transform(X_train)
X_test_processed = processor.transform(X_test)

eval_set = [(X_train_processed, y_train), 
            (X_test_processed, y_test)]

model = XGBClassifier(n_estimators=1000, n_jobs=-1)
model.fit(X_train_processed, y_train, eval_set=eval_set, eval_metric='auc', 
          early_stopping_rounds=10)
       

In [0]:
#2) Partial Dependence Plots with 1 feature

# Later, when you save matplotlib images to include in blog posts or web apps,
# increase the dots per inch (double it), so the text isn't so fuzzy
import matplotlib.pyplot as plt
plt.rcParams['figure.dpi'] = 72


In [0]:
#!pip list 

!pip install PDPbox

from pdpbox.pdp import pdp_isolate, pdp_plot
#from pdpbox.pdp import pdp, pdp_plot

feature = 'latitude'

isolated = pdp_isolate(
    model = RandomForestClassifier,
    dataset=X_train,
    model_features=X_train.columns,
    feature = feature
)


In [0]:
 #3) heat map of latitude, continent, and weather features

'''
X_train = X_train.pivot("latitude", "weather")
ax = sns.heatmap(X_train)
'''
# pandas pivot
heatmap1_data = pd.pivot_table(df3, values='latitude', 
                     index=['continent'], 
                     columns='weather')

sns.heatmap(heatmap1_data, cmap="YlGnBu")

In [0]:
#Try PDP https://colab.research.google.com/drive/1r2VFMtBAt3sLVIQfsMWyQXt8hB9gziRA#scrollTo=k5Tr9d1NcrCu (FINAL VERSION 234 u2s3m4.ipynb)

# chk target distribution/skew; Classes seem MOSTLY balanced Bar graph better for clsfn?
%matplotlib inline
import seaborn as sns
sns.distplot(y_train);

In [0]:
#PDP with Single Feature

target = 'appearedDayOfWeek'
features = df.columns.drop(['cellId_180m', 'cellId_370m'])

X = df3[features]
y = df3[target]

# Use Ordinal Encoder, outside of a pipeline
encoder = ce.OrdinalEncoder()
X_encoded = encoder.fit_transform(X)

# Fit a 2nd RandomForestClassifier, this time with random_state=42, n_jobs=-1 parameters, and
# using cellId_180m', 'cellId_370m' as features to compare (chosen from Corr Heat Map)

model = RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1)
# Fit a 2nd Rand
model = RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1)
model.fit(X_encoded, y)

In [0]:
# Accuracy of 2nd RandForCls model:  returns Test Accuracy: 0.6418263384049941 
print(f'Test Accuracy: {pipeline.score(X_test, y_test)}') # obivates #acc_score = pipeline.score(X_test, y_test) # runs pipeline ag X_Test and comps with actuals = y_test


In [0]:
# PDP PLOT Using Pdpbox
!pip install Pdpbox

%matplotlib inline
import matplotlib.pyplot as plt
from pdpbox import pdp
feature = 'latitude'
pdp_dist = pdp.pdp_isolate(model=model, dataset=X_encoded, model_features=features, feature=feature)
pdp.pdp_plot(pdp_dist, feature);

In [0]:
# SHAPLEY VALUES FORCE PLOT
!pip install shap
import shap

explainer = shap.TreeExplainer(model)
row_processed = processor.transform(row)
shap_values = explainer.shap_values(row_processed)

shap.initjs()
shap.force_plot(
    base_value=explainer.expected_value, 
    shap_values=shap_values, 
    features=row, 
    link='logit' # For classification, this shows predicted probabilities
)